hir-place-target.md

```diff
@@ -19,38 +19,44 @@ struct PlaceReference<'tcx> {
 struct Place<'tcx> {
     /// start of the place expression, typically a local variable
     base: PlaceBase,
-    
-    /// projections select parts of the base expression; e.g., 
+
+    /// Type of the Base
+    base_ty: Ty<'tcx>,
+
+    /// projections select parts of the base expression; e.g.,
     /// in the place expression `a.b.c`, `b` and `c` are projections
     projections: Vec<Projection<'tcx>>,
 }
 
-/// *Projections* select parts of the base expression; e.g., 
+/// *Projections* select parts of the base expression; e.g.,
 /// in the place expression `a.b.c`, `b` and `c` are projections
 struct Projection<'tcx> {
-    /// Type of the projection thus far.
-    ty: Ty<'tcx>,
-    
-    /// 
+    /// Type before the projection is applied.
+    before_ty: Ty<'tcx>,
```

Comment:
I don't think we should have *both* the before and after ty. We can compute the before-ty fairly easily after all.

Reply:
Can you point us to the code that computes the `before_ty`?  It wasn't quite obvious to me. More concretely, my concern is the code [here](https://github.com/rust-lang/rust/blob/master/src/librustc_typeck/mem_categorization.rs#L138) depends on the stored type information before projection is applied.

Reply:
It's not that there exists code *now* -- but you would have some code like this:

```rust
impl Place {
    fn ty_before_projection(&self, projection_index: usize) -> Ty<'tcx> {
        assert!(projection_index < self.projections.len());
        if projection_index == 0 { self.base_ty } else { self.projections[projection_index - 1].after_ty }
    }
}
```